### ISSUE TITLE
Non-concatenated DWI/EPI issue

### ISSUE THREAD

**Issue by araikes** [BIDS-Apps/MRtrix3_connectome#61]
Sorry for the slurry of issues but I figured it might be easier to have them separated out. Across all issues (#59, #60, and now this one), I have a Docker container that pulls the master for MRtrix3.

I rewrote the mrtrix_connectome.py file recently (see PR) and have further updated that today to stop concatenating the DWI and EPI files. Prior to this, `dwipreproc` did not have any issues. However, now it is throwing an error.

The script runs through, loads the DWI, EPI, and T1 images and converts them to .mifs as previously. Where it used to concatenate the DWI/EPI images, it now only concatenates multiple EPIs into `fmap_cat.mif`. 

Steps taken in processing:

1. DWI images are denoised and unringed, resulting in a file `dwipreproc_in.mif`
1. Field maps are concatenated if there are multiples and then unringed and saved as `se_epi.mif`
1. These are then fed to dwipreproc with the call: 
`dwipreproc dwipreproc_in.mif dwi_preprocessed.mif -rpe_header -se_epi se_epi.mif -align_seepi -eddy_options " --cnr_maps --repo" -eddyqc_all eddyqc/`
1. Inside of dwipreproc, `dwipreproc_in.mif` and `se_epi.mif` are copied to the temp file and headers are loaded. At the start of `dwipreproc` all files are present and accounted for.
1. I get the following output:
```
dwipreproc: Changing to temporary directory (/outputs/mrtrix3_connectome.py-tmp-9YA6MU/dwipreproc-tmp-6HU0HK/)
dwipreproc: Loading header for image file 'dwi.mif'
dwipreproc: Loading header for image file 'se_epi.mif'
dwipreproc: Command: '/mrtrix3/bin/mrinfo dwi.mif -shell_bvalues' (piping data to local storage)
dwipreproc: Result: 0 1000
Command:  dirstat dwi.mif -output asym
dwipreproc: DWIs and SE-EPI images used for inhomogeneity field estimation are defined on different image grids; the latter will be automatically re-gridded to match the former
Command:  mrtransform se_epi.mif - -interp sinc -template dwi.mif | mrcalc - 0.0 -max se_epi_regrid.mif
 dwipreproc: No phase-encoding contrast present in SE-EPI images; will examine again after combining with DWI b=0 images
Command:  dwiextract dwi.mif - -bzero | mrcat - se_epi.mif se_epi_regrid_dwibzeros.mif -axis 3
mrcat: [ERROR] failed to open key/value file "se_epi.mif": No such file or directory
mrcat: [ERROR] error opening image "se_epi.mif"
```
Based on lines [453-461](https://github.com/MRtrix3/mrtrix3/blob/81036fcc6dc11222515fc6cc1b2403585560bfcb/bin/dwipreproc#L453) of dwipreproc, it looks like when the regridding is done, the original se_epi.mif file is removed from the temp dwipreproc folder. However, with `-rpe_header` specified, line [519](https://github.com/MRtrix3/mrtrix3/blob/81036fcc6dc11222515fc6cc1b2403585560bfcb/bin/dwipreproc#L519) requires that file. I'm guessing that if regridding is done, then `se_epi.mif` should be replaced with `se_epi_regrid.mif`.

Because both the original se_epi image and the regridded version are assigned to `se_epi_path`, dwipreproc  line [519](https://github.com/MRtrix3/mrtrix3/blob/81036fcc6dc11222515fc6cc1b2403585560bfcb/bin/dwipreproc#L519) could probably be easily updated as:
`run.command('dwiextract dwi.mif - -bzero | mrcat - ' + se_epi_path + new_se_epi_path + ' -axis 3')`

**Comment by Lestropie** [comment]
> Sorry for the slurry of issues but I figured it might be easier to have them separated out.

Not at all; happy to have detailed feedback and for individual issues to be listed as ... individual issues. :upside_down_face: 

> Because both the original se_epi image and the regridded version are assigned to se_epi_path, dwipreproc line 519 could probably be easily updated as:

Yup, agree with your diagnosis there.

There's two 'specialised' `dwipreproc` functionalities going on in your particular data: *both* the `fmap` image(s) is/are on a different image grid to the DWIs, *and* you presumably have a single `fmap` image, which is not enough to estimate the inhomogeneity field and is why the script is trying to concatenate with the DWI *b*=0 images.

While the script should certainly be able to handle such data, and I'll make the upstream change, it is nevertheless worth confirming that this is in fact the nature of the data you have at hand. In order to reasonably be concatenated with the DWI *b*=0 images for estimation of the susceptibility field, the SE-EPI images need to have the same image *contrast* as the DWI *b*=0's. But if they are defined on a different voxel grid, *in addition to* not having intrinsic phase encoding contrast, my intuition suggests that this is more likely to be a *mis-use* of the script; for instance, attempting to use gradient-echo-based field map data rather than spin-echo EPI data, the former of which would almost certainly be on a different image voxel grid. Can you confirm that in your data, the image(s) in the `fmap/` directory is/are in fact spin-echo EPI, with comparable image contrast to the DWI *b*=0 images? The script is not yet clever enough to separate between dual-gradient-echo-based and spin-echo-EPI-based `fmap/` data; the latter will hopefully eventually be supported both here and upstream in `dwipreproc`, I've just never gotten around to it because I've never had the need to process such data myself.

**Comment by araikes** [comment]
Sorry, I think something may have been unclear about it when I wrote it. I left your code in place for concatenating the `fmap` images not because I need it but because someone else might. I've got a single RPE image to go along with the DWI we've collected. 

And yes, the `fmap` and `dwi` files are spin-echo EPIs with comparable contrasts. Can post screen shots if needed for further debugging this.

**Comment by Lestropie** [comment]
No that's fine, don't need evidence; just wanted to be sure that there wasn't some higher-level misunderstanding. :+1:

**Comment by araikes** [comment]
Dumb question that may go along with higher level misunderstanding...

I have a temp file with the `se_epi.mif` file that got created when I tried running the script (`se_epi_regridded.mif` is in the dwipreproc subfolder). If I `mrinfo` that file, there's a PhaseEncodingDirection keyval ("j") and one for the `dwipreproc_in.mif` file ("j-"). 

Should dwipreproc be picking that up for the phase encoding contrast?

**Comment by Lestropie** [comment]
What the script is doing in that area of the code is initially examining *just* the SE-EPI volumes to see if there is phase encoding contrast:

- If there is, then it will provide the SE-EPI volumes alone to `topup` for inhomogeneity field estimation.

- If however there is no such contrast, it takes into account the possibility that the acquisition may have been what you seem to have acquired: a single volume with the opposite phase encoding direction to the DWIs. So what it does in this case is concatenate the DWI *b*=0 volumes together with the SE-EPI image, and then check whether or not *that* concatenated series contains phase encoding contrast:

  - If it does, then the script can proceed by providing that concatenated data to `topup`;

  - If not, then it is impossible to estimate the susceptibility field using the image data provided to the script, and so it aborts.

What you should find with your data (and what the `dwipreproc` script actually explicitly uses to perform the check) is that if a 4D image series contains volumes where the phase encoding direction (and/or total readout time) is not equivalent across all volumes, then this gets stored in header key-value pair "`pe_scheme`" rather than "`PhaseEncodingDirection`" (and "`TotalReadoutTime`"). See documentation [here](https://mrtrix.readthedocs.io/en/latest/concepts/pe_scheme.html).

There's a fair amount of trickery involved in trying to have a script automatically process any and all acquisition protocol strategies :sweat:

**Comment by araikes** [comment]
Excellent. So it's doing what it should then in this case.... Once the regridded image gets used, in theory.

Any thoughts on why the script didn't have this issue previously when the dwi and fmap files were concatenated and split?

**Comment by Lestropie** [comment]
> Any thoughts on why the script didn't have this issue previously when the dwi and fmap files were concatenated and split?

I'd have expected the script to crash out at the initial `mrcat` call in the case where the DWI and `fmap` images don't have the same number of voxels in heac spatial axis; hence never getting to the point of exposing this issue.

**Comment by araikes** [comment]
Never had an issue before and it's run to completion for the preprocessing.

Odd too because both the dwi and fmap images are 128x128x60, so it seems like it shouldn't need regridding.

**Comment by Lestropie** [comment]
The internal logic by which two image voxel grids are determined to "match" - and hence voxel-wise operations (rather than scanner-space interpolation) are permitted - is different between the C++ binaries and the Python API. It's possible that there is a small floating-point difference between the two images that is deemed acceptable by `mrcat` but not acceptable by `dwipreproc`. I'm thinking specifically the header transformation might be slightly different due to a shift in FoV between the two acquisitions, and that might be adequate to trigger the `image.match()` function to return `False`.

**Comment by araikes** [comment]
That would make sense. The DWI loads in with a 2x2x2.0000000093 voxel size.

**Comment by Lestropie** [comment]
There should be [some tolerance](https://github.com/MRtrix3/mrtrix3/blob/3.0_RC3/lib/mrtrix3/image.py#L161-L165) in the voxel size comparison... Are you able to run `dwipreproc` with the `-debug` flag and go through the output to find out why it's classifying those two images as mismatched?

**Comment by araikes** [comment]
Log file attached
[logfile.txt](https://github.com/BIDS-Apps/MRtrix3_connectome/files/3648975/logfile.txt)

**Comment by Lestropie** [comment]
```
dwipreproc: [DEBUG] image.match() (from dwipreproc:454): 'dwi.mif' 'se_epi.mif' transform (rotation) mismatch (

[[0.998231001226161, 0.0519660786956019, 0.0288858891579189, -138.756545893476],
[-0.0561791997886395, 0.983466031365825, 0.172158245054531, -129.317351539818],
[-0.0194619038668969, -0.173476487785974, 0.984645694687841, 7.11091467716252],
[0.0, 0.0, 0.0, 1.0]] 

[[0.999582714972628, -1.51949994979406e-09, 0.0288859113076275, -132.37916280129],
[-0.00497502231774251, 0.985056744638102, 0.172158236250623, -135.969529145747],
[-0.0284542629651067, -0.172230107248952, 0.984645695577349, 8.08691801042225],
[0.0, 0.0, 0.0, 1.0]])
```

Okay. So the radiographer has used different FoVs in between the DWI acquisition, and acquisition of the single SE-EPI volume with reversed phase encoding. When it comes to voxel-wise processing, the magnitude of that discrepancy is adequately large that it would not make sense to loop through the two images voxel-by-voxel in conjunction to perform some processing task. However in the specific use case here, such a difference actually does not matter, since we're concatenating the data together in order to run a command that performs registration between volumes. Furthermore, the directionally-dependent information (the phase encoding direction) is fixed to the image axes rather than a real-space orientation, and thus is not corrupted by altering the header transformation associated with that image data.

So what you actually need here, in the ideal scenario, is for the `image.match()` function to ignore the image header transformations in this specific instance. That can be done; it just requires modification upstream in *MRtrix3*. `mrcat` will I think still produce a *warning* due to the transform mismatch, but the operation is technically permitted.

**Comment by Lestropie** [comment]
Should hopefully be resolved by upstream changes brought in by #62; can re-open if the issue is still encountered.
